### ISSUE TITLE
BO's tuning result doesn't provide the best hyperparameter configuration?

### ISSUE THREAD

**Issue by bblodfon** [mlr-org/mlr3mbo#84]
Hi,

I use an `AutoTuner` to train and tune a `learner` with a single optimization criterion. Compared to random search optimization, which outputs the `learner` with the hyperparameters that resulted in the best performance on the train data (in this case larger `surv.cindex`), `mbo` tuner doesn't, see below the corresponding `tuning_result`'s:
``` r
library(mlr3verse)
#> Loading required package: mlr3
library(mlr3proba)
library(xgboost)
library(tidyverse)
library(mlr3mbo)
#> Loading required package: mlr3tuning
#> Loading required package: paradox
library(survival)

# Less logging
lgr::get_logger('bbotk')$set_threshold('warn')
lgr::get_logger('mlr3')$set_threshold('warn')

set.seed(42)
train_indxs = sample(seq_len(nrow(veteran)), 100)
task = as_task_surv(x = veteran, time = 'time', event = 'status')
poe = po('encode')
task = poe$train(list(task))[[1]]
task
#> <TaskSurv:veteran> (137 x 11)
#> * Target: time, status
#> * Properties: -
#> * Features (9):
#>   - dbl (9): age, celltype.adeno, celltype.large, celltype.smallcell,
#>     celltype.squamous, diagtime, karno, prior, trt

ncores = 4
learner = lrn('surv.xgboost',
  nthread = ncores, booster = 'gbtree', early_stopping_rounds = 10,
  nrounds = to_tune(50, 1000),
  eta = to_tune(p_dbl(1e-04, 1, logscale = TRUE)),
  max_depth = to_tune(2, 10))

# Bayesian Optimization
xgboost_at_bo = AutoTuner$new(
  learner = learner,
  resampling = rsmp('cv', folds = 5),
  measure = msr('surv.cindex'),
  terminator = trm('evals', n_evals = 30),
  tuner = tnr('mbo')
)
xgboost_at_bo$train(task, row_ids = train_indxs)

xgboost_at_bo$tuning_result
#>    nrounds       eta max_depth learner_param_vals  x_domain surv.cindex
#> 1:     849 -6.270425         3          <list[7]> <list[3]>   0.6435471
xgboost_at_bo$archive$data %>%
  as_tibble() %>%
  arrange(desc(surv.cindex)) %>% print(n = 5)
#> # A tibble: 30 × 12
#>   nrounds    eta max_depth surv.cindex warni…¹ errors runti…² uhash x_domain    
#>     <int>  <dbl>     <int>       <dbl>   <int>  <int>   <dbl> <chr> <list>      
#> 1     623 -6.38          5       0.662       0      0   1.90  8cc2… <named list>
#> 2     361 -0.640         9       0.655       0      0   0.263 142b… <named list>
#> 3     769 -6.26          5       0.650       0      0   2.30  38e7… <named list>
#> 4     523 -6.42          5       0.649       0      0   1.73  1e3c… <named list>
#> 5     443 -6.39          2       0.648       0      0   1.31  a962… <named list>
#> # … with 25 more rows, 3 more variables: timestamp <dttm>, batch_nr <int>,
#> #   acq_ei <dbl>, and abbreviated variable names ¹​warnings, ²​runtime_learners
#> # ℹ Use `print(n = ...)` to see more rows, and `colnames()` to see all variable names

# Random Search
xgboost_at_rs = AutoTuner$new(
  learner = learner,
  resampling = rsmp('cv', folds = 5),
  measure = msr('surv.cindex'),
  terminator = trm('evals', n_evals = 30),
  tuner = tnr('random_search')
)
xgboost_at_rs$train(task, row_ids = train_indxs)

xgboost_at_rs$tuning_result
#>    nrounds      eta max_depth learner_param_vals  x_domain surv.cindex
#> 1:     311 -8.61813         3          <list[7]> <list[3]>   0.7027073
xgboost_at_rs$archive$data %>%
  as_tibble() %>%
  arrange(desc(surv.cindex)) %>% print(n = 5)
#> # A tibble: 30 × 11
#>   nrounds   eta max_depth surv.cindex warnings errors runti…¹ uhash x_domain    
#>     <int> <dbl>     <int>       <dbl>    <int>  <int>   <dbl> <chr> <list>      
#> 1     311 -8.62         3       0.703        0      0   1.79  a525… <named list>
#> 2     657 -7.35         4       0.693        0      0   2.25  cd38… <named list>
#> 3     163 -4.95         2       0.685        0      0   0.514 3084… <named list>
#> 4     184 -5.22         4       0.674        0      0   0.656 72d5… <named list>
#> 5     162 -5.16         5       0.672        0      0   0.639 c6a6… <named list>
#> # … with 25 more rows, 2 more variables: timestamp <dttm>, batch_nr <int>, and
#> #   abbreviated variable name ¹​runtime_learners
#> # ℹ Use `print(n = ...)` to see more rows, and `colnames()` to see all variable names
```

<sup>Created on 2022-08-29 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

----

So, is this a bug or a feature (e.g. `mbo` does something else smart under the 'hood' to choose the best hyperparameter configuration)?

**Comment by bblodfon** [comment]
Thanks @sumny , now it makes sense :) Some more things that popped up:
1) I installed the latest `mlr3mbo` version from GitHub, but:
```
default_result = function (inst, optimizer_mbo) {
  res = inst$archive$best()
  xdt = res[, inst$search_space$ids(), with = FALSE]
  y = unlist(res[, inst$archive$cols_y, with = FALSE])
  inst$assign_result(xdt, y)
  invisible(NULL)
}
tuner$result_function = default_result
Error in checkFunction(x, args, ordered, nargs, null.ok) : 
  object 'result_function' not found
```
2) Is there some quick way to get the `xgboost_at_bo$archive$best()` parameters and assigned them to the output final learner, so that I retrain it manually using those?
3) Why do the mean CV `surv.cindex` is **different** between `xgboost_at_bo$tuning_result` and `xgboost_at_bo$archive$best()`? (though hyperparams are the same!). See below:

``` r
library(mlr3verse)
library(mlr3proba)
library(xgboost)
library(tidyverse)
library(mlr3mbo)
library(survival)

# Less logging
lgr::get_logger('bbotk')$set_threshold('warn')
lgr::get_logger('mlr3')$set_threshold('warn')

set.seed(42)
train_indxs = sample(seq_len(nrow(veteran)), 100)
test_indxs = setdiff(seq_len(nrow(veteran)), train_indxs)
task = as_task_surv(x = veteran, time = 'time', event = 'status')
poe = po('encode')
task = poe$train(list(task))[[1]]
task
#> <TaskSurv:veteran> (137 x 11)
#> * Target: time, status
#> * Properties: -
#> * Features (9):
#>   - dbl (9): age, celltype.adeno, celltype.large, celltype.smallcell,
#>     celltype.squamous, diagtime, karno, prior, trt

ncores = 4
learner = lrn('surv.xgboost',
  nthread = ncores, booster = 'gbtree',
  nrounds = to_tune(50, 1000),
  eta = to_tune(p_dbl(1e-04, 1, logscale = TRUE)),
  max_depth = to_tune(2, 10))

# Bayesian Optimization ----
xgboost_at_bo = AutoTuner$new(
  learner = learner,
  resampling = rsmp('cv', folds = 5),
  measure = msr('surv.cindex'),
  terminator = trm('evals', n_evals = 30),
  tuner = tnr('mbo')
)
xgboost_at_bo$train(task, row_ids = train_indxs)

xgboost_at_bo$tuning_result
#>    nrounds       eta max_depth learner_param_vals  x_domain surv.cindex
#> 1:     623 -6.383083         5          <list[6]> <list[3]>   0.6548733
xgboost_at_bo$archive$best()
#>    nrounds       eta max_depth surv.cindex warnings errors runtime_learners
#> 1:     623 -6.383083         5   0.6617413        0      0            1.461
#>                                   uhash  x_domain           timestamp batch_nr
#> 1: 50720064-4944-4d2c-bf6c-770c186367ae <list[3]> 2022-08-29 15:45:14        1
#>    acq_ei
#> 1:     NA

# seems to be the second best `0.654` that is reported in the `tuning_result`?
xgboost_at_bo$archive$data %>%
  as_tibble() %>%
  select(nrounds, eta, max_depth, surv.cindex, uhash) %>%
  arrange(desc(surv.cindex)) %>% # note that the surv.cindex is the CV average
  slice(n = 1:5)
#> # A tibble: 5 × 5
#>   nrounds   eta max_depth surv.cindex uhash                               
#>     <int> <dbl>     <int>       <dbl> <chr>                               
#> 1     623 -6.38         5       0.662 50720064-4944-4d2c-bf6c-770c186367ae
#> 2     645 -5.88         5       0.654 db61fa9a-e090-4cd2-97ae-8fcf63301b7a
#> 3      53 -4.56         2       0.651 a6a721c1-cf6c-41dc-84d5-aefc77308343
#> 4     769 -6.26         5       0.650 73e90e87-9df2-4de5-9b9b-abd222c54365
#> 5     125 -5.65         2       0.649 cc8e8db5-5dc9-4a7b-b7e8-b43d0b6a7da3
```

<sup>Created on 2022-08-29 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

**Comment by sumny** [comment]
1. When did you install the GitHub Version?
The fix was just pushed today: https://github.com/mlr-org/mlr3mbo/commit/961ed7aa875382406bd22cb00d3080110ddc1ab6

2. You can access this hyperparameter configuration as a list via `xgboost_at_bo$archive$best()$x_domain[[1L]]`, e.g.:
```
best_hpc = xgboost_at_bo$archive$best()$x_domain[[1L]]
learner$param_set$values = insert_named(learner$param_set$values, best_hpc)
```

4. That was actually a bug. The `surv.cindex` in the `tuning_result` was actually the prediction of the surrogate.
I fixed this here: https://github.com/mlr-org/mlr3mbo/commit/d1159e293c3674e06bd57ca81f1d19d7aa9aea95
Thanks for reporting it! :)
On a sidenote: Please be aware that this performance is not an unbiased estimate of the performance of the learner configured by the optimal (based on the tuning result) hyperparameters. To get an unbiased performance estimate, you should perform nested resampling (essentially wrapping the `AutoTuner` in an outer resampling, see, https://mlr3book.mlr-org.com/optimization.html#nested-resampling)

**Comment by bblodfon** [comment]
1. I reinstalled the package now and it works, who knows what happened some hours ago!
2. Thanks!
3. (or 4 :) => glad to be of help! I am aware of the bias you mention and the nested-CV as a potential solution - learned it actually from reading the mlr3 book (you guys are doing **awesome** **work**)! In this example, I have splitted to train and test sets and I am using the best learner out of the tuning on the test set + some bootstrapping to get confidence intervals which has the advantage of being much faster than performaing nested-CV on the datasets I am working on.
----
A design consideration: would it make sense to have an initialization option in the tuner to use the `default_result` function (`archive$best()`) without the user having to write it themselves? Something like:
```
tnr('mbo', result = 'best')
tnr('mbo', result = 'mean_surrogate') # default
```
